crates/next-core/src/next_shared/webpack_rules/sass.rs

```diff
@@ -1,24 +1,76 @@
-use std::mem::take;
+use std::{io::Read, mem::take};
 
 use anyhow::{bail, Result};
+use semver::Version;
 use serde_json::Value as JsonValue;
 use turbo_tasks::{ResolvedVc, Vc};
+use turbo_tasks_fs::{FileContent, FileSystemPath};
 use turbopack::module_options::{LoaderRuleItem, OptionWebpackRules, WebpackRules};
+use turbopack_core::{
+    asset::{Asset, AssetContent},
+    resolve::{pattern::Pattern, resolve_raw},
+};
 use turbopack_node::transforms::webpack::WebpackLoaderItem;
 
 #[turbo_tasks::function]
 pub async fn maybe_add_sass_loader(
     sass_options: Vc<JsonValue>,
     webpack_rules: Option<Vc<WebpackRules>>,
+    package_dir: ResolvedVc<FileSystemPath>,
 ) -> Result<Vc<OptionWebpackRules>> {
     let sass_options = sass_options.await?;
     let Some(mut sass_options) = sass_options.as_object().cloned() else {
         bail!("sass_options must be an object");
     };
-    // TODO: Remove this once we upgrade to sass-loader 16
+
+    let sass_package_json_path = resolve_raw(
+        *package_dir,
+        Pattern::Constant("sass/package.json".into()).cell(),
+        false,
+    )
+    .first_source()
+    .await?;
+
+    let sass_version = match &*sass_package_json_path {
+        Some(sass_package_json) => {
+            let sass_package_json_content = sass_package_json.content().await?;
+            match &*sass_package_json_content {
+                AssetContent::File(file_content) => match &*file_content.await? {
+                    FileContent::Content(file) => {
+                        let mut reader = file.read();
+                        let mut buf = Vec::new();
+                        reader.read_to_end(&mut buf)?;
+                        let json_value = serde_json::from_slice::<serde_json::Value>(&buf)?;
+                        json_value
+                            .get("version")
+                            .unwrap()
+                            .as_str()
+                            .map(Version::parse)
+                    }
+                    _ => None,
+                },
+                _ => None,
+            }
+        }
+        None => None,
+    }
+    .transpose()?
+    .unwrap_or_else(|| Version::new(1, 45, 0));
+
+    // The modern Sass API with breaking changes was added in sass@1.45.0.
+    // https://sass-lang.com/documentation/breaking-changes/legacy-js-api
+    // Since sass-loader and our peer dependency sass version is ^1.3.0,
+    // we need to use the legacy Sass API for versions less than 1.45.0.
+    let should_use_legacy_sass_api = sass_version < Version::parse("1.45.0")?;
+
+    // TODO(jiwon): Once the peer dependency for sass is upgraded to >= 1.45.0, we can remove this.
     sass_options.insert(
-        "silenceDeprecations".into(),
-        serde_json::json!(["legacy-js-api"]),
+        "api".into(),
+        serde_json::json!(if should_use_legacy_sass_api {
+            "legacy"
+        } else {
+            "modern"
+        }),
```

Comment:
`"modern-compiler"` for sass-embedded

https://nextjs.org/docs/app/building-your-application/styling/sass#implementation
https://vite.dev/config/shared-options#css-preprocessoroptions